src/Lib.P2P/PeerConnection.cs

```diff
@@ -0,0 +1,415 @@
+            Stream stream,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            var protocols = ProtocolRegistry.Protocols.Keys
+               .Where(k => k == name || k.StartsWith(name))
+               .Select(k => VersionedName.Parse(k))
+               .OrderByDescending(vn => vn)
+               .Select(vn => vn.ToString());
+            foreach (var protocol in protocols)
+            {
+                await Message.WriteAsync(protocol, stream, cancel).ConfigureAwait(false);
+                var result = await Message.ReadStringAsync(stream, cancel).ConfigureAwait(false);
+                if (result == protocol) return;
+            }
+
+            if (protocols.Count() == 0) throw new Exception($"Protocol '{name}' is not registered.");
+            throw new Exception($"{RemotePeer.Id} does not support protocol '{name}'.");
+        }
+
+        /// <summary>
+        ///   Starts reading messages from the remote peer.
+        /// </summary>
+        public async Task ReadMessagesAsync(CancellationToken cancel)
+        {
+            log.Debug($"start reading messsages from {RemoteAddress}");
+
+            // TODO: Only a subset of protocols are allowed until
+            // the remote is authenticated.
+            IPeerProtocol protocol = new Multistream1();
+            try
+            {
+                while (!cancel.IsCancellationRequested && Stream != null)
+                    await protocol.ProcessMessageAsync(this, Stream, cancel).ConfigureAwait(false);
+            }
+            catch (IOException e)
+            {
+                log.Error("reading message failed " + e.Message);
+
+                // eat it.
+            }
+            catch (Exception e)
+            {
+                if (!cancel.IsCancellationRequested && Stream != null) log.Error("reading message failed", e);
+            }
+
+            // Ignore any disposal exceptions.
+            try
+            {
+                Stream?.DisposeAsync();
+            }
+            catch (Exception)
+            {
+                // eat it.
+            }
+
+            log.Debug($"stop reading messsages from {RemoteAddress}");
+        }
+
+        /// <summary>
+        ///   Starts reading messages from the remote peer on the specified stream.
+        /// </summary>
+        public async Task ReadMessagesAsync(Stream stream, CancellationToken cancel)
+        {
+            IPeerProtocol protocol = new Multistream1();
+            try
+            {
+                while (!cancel.IsCancellationRequested && stream != null && stream.CanRead)
+                    await protocol.ProcessMessageAsync(this, stream, cancel).ConfigureAwait(false);
+            }
+            catch (EndOfStreamException)
+            {
+                // eat it.
+            }
+            catch (Exception e)
+            {
+                if (!cancel.IsCancellationRequested && stream != null)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)